internal/job/shell/shell.go

```diff
@@ -269,6 +274,48 @@ func (s *Shell) RunWithEnv(ctx context.Context, environ *env.Environment, comman
 	})
 }
 
+// RunWithOlfactor runs a command, writes stdout and stderr to the shell's writer,
+// and returns an error if it fails. If the process exits with a non-zero exit code,
+// and `smell` was written to the logger (i.e. the combined stream of stdout and stderr),
+// the error will be of type `olfactor.OlfactoryError`. If the process exits 0, the error
+// will be nil.
+func (s *Shell) RunWithOlfactor(
+	ctx context.Context,
+	smell string,
+	command string,
+	arg ...string,
+) error {
+	formatted := process.FormatCommand(command, arg)
+	if s.stdin == nil {
+		s.Promptf("%s", formatted)
+	} else {
+		// bash-syntax-compatible indication that input is coming from somewhere
+		s.Promptf("%s < /dev/stdin", formatted)
+	}
+
+	cmd, err := s.buildCommand(command, arg...)
+	if err != nil {
+		s.Errorf("Error building command: %v", err)
+		return err
+	}
+
+	w, o := olfactor.New(io.Discard, smell)
+	mw := io.MultiWriter(s.Writer, w)
```

Comment:
[starting debates where none are needed] Although there is no need for a dependency between `s.Writer` and `w`, could the `MultiWriter` be eliminated by passing `s.Writer` into `olfactor.New`?

Reply:
Whoops, this is much clearer. The dependency is a good thing, I think. It means they are part of a path in a pipeline.